src/main/java/com/uber/cadence/internal/sync/ActivityExecutionContextImpl.java

```diff
@@ -45,28 +49,88 @@
   private static final Logger log = LoggerFactory.getLogger(ActivityExecutionContextImpl.class);
 
   private final IWorkflowService service;
-
   private final String domain;
-
   private final ActivityTask task;
   private final DataConverter dataConverter;
   private boolean doNotCompleteOnReturn;
+  private final long heartbeatIntervalNanos;
+  private Object lastDetails;
+  private final ScheduledExecutorService heartbeatExecutor;
+  private Lock lock = new ReentrantLock();
+  private ScheduledFuture future;
+  private ActivityCompletionException lastException;
 
   /** Create an ActivityExecutionContextImpl with the given attributes. */
   ActivityExecutionContextImpl(
-      IWorkflowService service, String domain, ActivityTask task, DataConverter dataConverter) {
+      IWorkflowService service,
+      String domain,
+      ActivityTask task,
+      DataConverter dataConverter,
+      ScheduledExecutorService heartbeatExecutor) {
     this.domain = domain;
     this.service = service;
     this.task = task;
     this.dataConverter = dataConverter;
+    this.heartbeatIntervalNanos = (long) (0.8 * task.getHeartbeatTimeout().toNanos());
+    this.heartbeatExecutor = heartbeatExecutor;
   }
 
-  /**
-   * @throws CancellationException
-   * @see ActivityExecutionContext#recordActivityHeartbeat(Object)
-   */
+  /** @see ActivityExecutionContext#recordActivityHeartbeat(Object) */
   @Override
   public void recordActivityHeartbeat(Object details) throws ActivityCompletionException {
```

Comment:
details can be null. I think logic is broken in this case.

Reply:
Good catch. I fixed the logic and add a new unit test to cover this logic.